core/src/leader_schedule_cache.rs

```diff
@@ -0,0 +1,98 @@
+use crate::leader_schedule::LeaderSchedule;
+use crate::leader_schedule_utils;
+use solana_runtime::bank::{Bank, EpochSchedule};
+use solana_sdk::pubkey::Pubkey;
+use std::collections::HashMap;
+use std::sync::RwLock;
+
+const MAX_SCHEDULES: usize = 10;
+
+pub struct LeaderScheduleCache {
+    // Map from an epoch to a leader schedule for that epoch
+    pub cached_schedules: RwLock<HashMap<u64, LeaderSchedule>>,
+    epoch_schedule: EpochSchedule,
+}
+
+impl LeaderScheduleCache {
+    pub fn new(epoch_schedule: EpochSchedule) -> Self {
+        Self {
+            cached_schedules: RwLock::new(HashMap::new()),
+            epoch_schedule,
+        }
+    }
+
+    pub fn slot_leader_at(&self, slot: u64) -> Option<Pubkey> {
+        let (epoch, slot_index) = self.epoch_schedule.get_epoch_and_slot_index(slot);
+        self.cached_schedules
+            .read()
+            .unwrap()
+            .get(&epoch)
+            .map(|schedule| schedule[slot_index])
+    }
+
+    pub fn slot_leader_at_else_compute(&self, bank: &Bank) -> Option<Pubkey> {
+        let cache_result = self.slot_leader_at(bank.slot());
+        if cache_result.is_some() {
+            cache_result
+        } else {
+            let (epoch, slot_index) = bank.get_epoch_and_slot_index(bank.slot());
+            let leader_schedule = leader_schedule_utils::leader_schedule(epoch, bank);
+            leader_schedule.map(|leader_schedule| {
+                let leader = leader_schedule[slot_index];
+                let mut cached_schedules = self.cached_schedules.write().unwrap();
+                cached_schedules.insert(epoch, leader_schedule);
+                Self::retain_latest(&mut cached_schedules);
+                leader
+            })
+        }
+    }
+
+    fn retain_latest(schedules: &mut HashMap<u64, LeaderSchedule>) {
+        if schedules.len() > MAX_SCHEDULES {
```

Comment:
the Bank never culls epoch_vote_accounts, might not really be necessary for you to do any culling for a *long* time

Reply:
200 schedules? roughly a year?

Reply:
yeah, like noise in the scheme of things...